### PR TITLE
Fix connection leak when add_pool is called in rapid succession

### PR DESCRIPTION
**Pull request by huss01**

### CODE REVIEW

File: src/emysql.erl [hunk 1]
```diff
@@ -293,7 +293,15 @@ add_pool(#pool{pool_id=PoolId,size=Size,user=User,password=Password,host=Host,po
...
+                    ok;
+                {'EXIT', pool_already_exists} ->
+                    emysql_conn:close_connections(Pool2),
+                    {error, pool_already_exists};
+                {'EXIT', Other} ->
+                    exit(Other)
+            end
     end.
 
 %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result
```
Comment (elvisci) [inline src/emysql.erl]: Line 307 is too long: %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result.

File: src/emysql_conn.erl [hunk 2]
```diff
@@ -305,6 +301,13 @@ reset_connection(Pools, Conn, StayLocked) ->
...
+close_connections(Pool) ->
+    AllConns = lists:append(
+                 queue:to_list(Pool#pool.available),
+                 gb_trees:values(Pool#pool.locked)
+                ),
+    lists:foreach(fun emysql_conn:close_connection/1, AllConns).
+
 close_connection(Conn) ->
 	%% garbage collect statements
 	emysql_statements:remove(Conn#emysql_connection.id),
```
Comment (elvisci) [inline src/emysql_conn.erl]: Line 313 has a tab at column 0.

File: src/emysql_conn.erl [hunk 3]
```diff
@@ -171,13 +171,9 @@ open_connections(Pool) ->
                 #emysql_connection{} = Conn ->
                     open_connections(Pool#pool{available = queue:in(Conn, Pool#pool.available)});
```
Comment (elvisci) [inline src/emysql_conn.erl]: Line 172 is too long:                     open_connections(Pool#pool{available = queue:in(Conn, Pool#pool.available)});.

File: src/emysql_conn.erl [hunk 4]
```diff
@@ -305,6 +301,13 @@ reset_connection(Pools, Conn, StayLocked) ->
...
 
+close_connections(Pool) ->
+    AllConns = lists:append(
+                 queue:to_list(Pool#pool.available),
+                 gb_trees:values(Pool#pool.locked)
+                ),
+    lists:foreach(fun emysql_conn:close_connection/1, AllConns).
+
 close_connection(Conn) ->
 	%% garbage collect statements
```
Comment (elvisci) [inline src/emysql_conn.erl]: Line 312 has a tab at column 0.

File: src/emysql.erl [hunk 5]
```diff
@@ -293,7 +293,15 @@ add_pool(#pool{pool_id=PoolId,size=Size,user=User,password=Password,host=Host,po
                 {ok, Pool1} -> Pool1;
                 {error, Reason} -> throw(Reason)
             end,
-            emysql_conn_mgr:add_pool(Pool2)
+            case catch emysql_conn_mgr:add_pool(Pool2) of
```
Comment (elbrujohalcon) [inline src/emysql.erl]: would you mind using `try…catch` instead of `case catch`?

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: src/emysql.erl [hunk 6]
```diff
@@ -293,7 +293,12 @@ add_pool(#pool{pool_id=PoolId,size=Size,user=User,password=Password,host=Host,po
...
-            emysql_conn_mgr:add_pool(Pool2)
+            try
+                emysql_conn_mgr:add_pool(Pool2)
+            catch
+                exit:pool_already_exists ->
+                    {error, pool_already_exists}
+            end
     end.
 
 %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result
```
Comment (elvisci) [inline src/emysql.erl]: Line 304 is too long: %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result.

**Review by elbrujohalcon — APPROVED** [review]

File: src/emysql_conn.erl (hunk 2 above)
Comment (elvisci) [inline src/emysql_conn.erl]: Line 313 has a tab at column 0.

File: src/emysql_conn.erl (hunk 3 above)
Comment (elvisci) [inline src/emysql_conn.erl]: Line 172 is too long:                     open_connections(Pool#pool{available = queue:in(Conn, Pool#pool.available)});.

File: src/emysql_conn.erl (hunk 4 above)
Comment (elvisci) [inline src/emysql_conn.erl]: Line 312 has a tab at column 0.

File: src/emysql.erl [hunk 7]
```diff
@@ -293,7 +293,13 @@ add_pool(#pool{pool_id=PoolId,size=Size,user=User,password=Password,host=Host,po
...
+            try
+                emysql_conn_mgr:add_pool(Pool2)
+            catch
+                exit:pool_already_exists ->
+                    emysql_conn:close_connections(Pool2),
+                    {error, pool_already_exists}
+            end
     end.
 
 %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result
```
Comment (elvisci) [inline src/emysql.erl]: Line 305 is too long: %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result.

File: src/emysql_conn.erl (hunk 2 above)
Comment (elvisci) [inline src/emysql_conn.erl:313]: Line 313 has a tab at column 0.

File: src/emysql.erl (hunk 7 above)
Comment (elvisci) [inline src/emysql.erl:305]: Line 305 is too long: %% @spec add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding) -> Result.

File: src/emysql_conn.erl (hunk 4 above)
Comment (elvisci) [inline src/emysql_conn.erl:312]: Line 312 has a tab at column 0.

File: src/emysql_conn.erl (hunk 3 above)
Comment (elvisci) [inline src/emysql_conn.erl:172]: Line 172 is too long:                     open_connections(Pool#pool{available = queue:in(Conn, Pool#pool.available)});.

**Pull request merged**